builder/assignmentEnv.go

```diff
@@ -0,0 +1,315 @@
+		}
+		defer func() {
+			err = response.Body.Close()
+			if err != nil {
+				log.Println(err)
+				return
+			}
+		}()
+
+		_, err = io.Copy(os.Stdout, response.Body)
+		if err != nil {
+			return errors.Wrap(err, "error in reading image build response")
+		}
+	}
+	return nil
+}
+
+func (assgnEnv *assignmentEnv) publish() error {
+	if assgnEnv.DockerfileData.Len() <= 0 {
+		return assgnEnv.pullImage()
+	} else if assgnEnv.ImgBuildConfig.publishImage {
+		return assgnEnv.pushImage()
+	}
+	return nil
+}
+
+func (assgnEnv *assignmentEnv) pullImage() error {
+	backgroundContext := context.Background()
+	dockerClient, err := client.NewEnvClient()
+	if err != nil {
+		return errors.Wrap(err, "error in creating new docker client")
+	}
+	authConfig := types.AuthConfig{
+		Username: assgnEnv.ImgBuildConfig.authData.Username,
+		Password: assgnEnv.ImgBuildConfig.authData.Password,
+	}
+	encodedJSON, err := json.Marshal(authConfig)
+	if err != nil {
+		return errors.Wrap(err, "error in encoding authConfig")
+	}
+
+	authStr := base64.URLEncoding.EncodeToString(encodedJSON)
+
+	resp, err := dockerClient.ImagePull(backgroundContext, assgnEnv.ImgBuildConfig.imageTag, types.ImagePullOptions{
+		RegistryAuth: authStr,
+	})
+
+	if err != nil {
+		return errors.Wrap(err, "error in pulling image from hub")
+	}
+	_, err = io.Copy(os.Stdout, resp)
+	if err != nil {
+		return errors.Wrap(err, "error in reading image pull response")
+	}
+	defer func() {
+		err = resp.Close()
+		if err != nil {
+			log.Println(err)
+			return
+		}
+	}()
+
+	return nil
+
+}
+func (assgnEnv *assignmentEnv) pushImage() error {
+	backgroundContext := context.Background()
+	dockerClient, err := client.NewEnvClient()
+	if err != nil {
+		return errors.Wrap(err, "error in creating new docker client")
+	}
+
+	authConfig := types.AuthConfig{
+		Username: assgnEnv.ImgBuildConfig.authData.Username,
+		Password: assgnEnv.ImgBuildConfig.authData.Password,
+	}
+	encodedJSON, err := json.Marshal(authConfig)
+	if err != nil {
+		return errors.Wrap(err, "error in encoding authConfig")
+	}
+
+	authStr := base64.URLEncoding.EncodeToString(encodedJSON)
+
+	imageString := fmt.Sprintf("%s/%s", constants.DockerIO, assgnEnv.ImgBuildConfig.imageTag)
+
+	resp, err := dockerClient.ImagePush(backgroundContext, imageString, types.ImagePushOptions{
+		RegistryAuth: authStr,
+	})
+	if err != nil {
+		return errors.Wrap(err, "error in pushing image to hub")
+	}
+	_, err = io.Copy(os.Stdout, resp)
+	if err != nil {
+		return errors.Wrap(err, "error in reading image push response")
+	}
+	defer func() {
+		err = resp.Close()
+		if err != nil {
+			log.Println(err)
+			return
+		}
+	}()
+	return nil
+}
+
+func (assgnEnv *assignmentEnv) undoWrite() {
+	// Clear the dockerfile data bytes.
+	assgnEnv.DockerfileData.Reset()
+}
+
+func (assgnEnv *assignmentEnv) undoBuild() error {
+	// Delete the created Dockerfile.
+	return os.Remove(assgnEnv.ImgBuildConfig.dockerFilepath)
```

Comment:
how is deleting the Dockerfile going to undo the build?
Deleting the Dockerfile doesn't get rid of the image built.

Reply:
Oh no no, I will rename the undo functions, this function is actually undoing the dockerfile write operation, whereas the built image will be deleted if image push fails.